In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
# __

import geopandas as gpd
import geoplot as gplt
import geoplot.crs as gcrs
from shapely.geometry import Point
# __

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler, MinMaxScaler

# Random Forest resisting on overfit, capture non linear tedencies in the data
from sklearn.ensemble import RandomForestClassifier

# Logistic model
from sklearn.linear_model import LogisticRegression

# To upgrape and estimated the model to have a better score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import precision_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score

import statsmodels.api as sm
from sklearn.model_selection import train_test_split

# __
from pivottablejs import pivot_ui

[source](https://www.kaggle.com/datasets/anoopjohny/real-estate-sales-2001-2020-state-of-connecticut)
- Serial Number: A unique identifier for each transaction, facilitating easy tracking and reference throughout the entire 20-year
- List Year: Specifies the year to which each transaction is associated, enabling historical analysis and trend identification -> **officially put up for sale**
- Date Recorded: The date when each transaction was officially recorded, providing a chronological view of real estate sales activity -> **date sold**
- Town: Identifies the specific town within Connecticut where each property is located, allowing for regional and temporal
- Address: The property address for each transaction, offering a detailed inventory of properties sold
- Assessed Value: The assessed value of each property, a critical component for calculating property taxes
- Sale Amount: The actual sale price or transaction amount, serving as a primary indicator of property market value
- Sales Ratio: The ratio of the sale amount to the assessed value, useful for assessing the fairness and accuracy of property
- Property Type: Categorizes properties into types such as Residential, Commercial, Industrial, or others, offering insights into the ...
- Residential Type: Provides further classification for residential properties, including Single Family, Condo, or Two Family, ...
- Non Use Code: A code that may denote special use or classification of the property, if applicable
- Assessor Remarks: Notes or comments provided by property assessors, adding context and historical insights to each ...
- OPM remarks: Remarks or notes provided by the Office of Policy and Management, potentially containing administrative or ... 
- Location: Geographic coordinates (latitude and longitude) of each property, enabling spatial and temporal analyses, as ...

In [2]:
#df = pd.read_csv("/Users/jannikherrlich/Downloads/RES.csv")
df = pd.read_csv("RES.csv")

C:\Users\eric-\AppData\Local\Temp\ipykernel_101516\787604014.py:2: DtypeWarning: Columns (8,9,10,11,12) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("RES.csv")


# Exploratory Data Analysis (EDA)

In [3]:
df

Serial Number  List Year Date Recorded           Town  \
0               220008       2022    01/30/2023        Andover   
1              2020348       2020    09/13/2021        Ansonia   
2                20002       2020    10/02/2020        Ashford   
3               210317       2021    07/05/2022           Avon   
4               200212       2020    03/09/2021           Avon   
...                ...        ...           ...            ...   
1097624         190272       2019    06/24/2020     New London   
1097625         190284       2019    11/27/2019      Waterbury   
1097626         190129       2019    04/27/2020  Windsor Locks   
1097627         190504       2019    06/03/2020     Middletown   
1097628         190344       2019    12/20/2019        Milford   

                  Address  Assessed Value  Sale Amount  Sales Ratio  \
0             618 ROUTE 6        139020.0     232000.0     0.599200   
1         230 WAKELEE AVE        150500.0     325000.0     0.463000   
2         390 TURNPIKE RD        253000.0     430000.0     0.588300   
3         53 COTSWOLD WAY        329730.0     805000.0     0.409600   
4        5 CHESTNUT DRIVE        130400.0     179900.0     0.724800   
...                   ...             ...          ...          ...   
1097624       4 BISHOP CT         60410.0      53100.0     1.137665   
1097625   126 PERKINS AVE         68280.0      76000.0     0.898400   
1097626    19 HATHAWAY ST        121450.0     210000.0     0.578300   
1097627      8 BYSTREK DR        203360.0     280000.0     0.726300   
1097628   250 RESEARCH DR       4035970.0    7450000.0     0.541700   

         Property Type Residential Type      Non Use Code Assessor Remarks  \
0          Residential    Single Family               NaN              NaN   
1           Commercial              NaN               NaN              NaN   
2          Residential    Single Family               NaN              NaN   
3          Residential    Single Family               NaN              NaN   
4          Residential            Condo               NaN              NaN   
...                ...              ...               ...              ...   
1097624  Single Family    Single Family  14 - Foreclosure              NaN   
1097625  Single Family    Single Family        25 - Other     PRIVATE SALE   
1097626  Single Family    Single Family               NaN              NaN   
1097627  Single Family    Single Family               NaN              NaN   
1097628            NaN              NaN               NaN              NaN   

        OPM remarks                            Location  
0               NaN  POINT (-72.343628962 41.728431984)  
1               NaN                                 NaN  
2               NaN                                 NaN  
3               NaN  POINT (-72.846365959 41.781677018)  
4               NaN                                 NaN  
...             ...                                 ...  
1097624         NaN                                 NaN  
1097625         NaN                                 NaN  
1097626         NaN                                 NaN  
1097627         NaN                                 NaN  
1097628         NaN                                 NaN  

[1097629 rows x 14 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1097629 entries, 0 to 1097628
Data columns (total 14 columns):
 #   Column            Non-Null Count    Dtype  
---  ------            --------------    -----  
 0   Serial Number     1097629 non-null  int64  
 1   List Year         1097629 non-null  int64  
 2   Date Recorded     1097627 non-null  object 
 3   Town              1097629 non-null  object 
 4   Address           1097578 non-null  object 
 5   Assessed Value    1097629 non-null  float64
 6   Sale Amount       1097629 non-null  float64
 7   Sales Ratio       1097629 non-null  float64
 8   Property Type     715183 non-null   object 
 9   Residential Type  699240 non-null   object 
 10  Non Use Code      313451 non-null   object 
 11  Assessor Remarks  171228 non-null   object 
 12  OPM remarks       13031 non-null    object 
 13  Location          298111 non-null   object 
dtypes: float64(3), int64(2), object(9)
memory usage: 117.2+ MB


In [5]:
df.describe()

Serial Number     List Year  Assessed Value   Sale Amount   Sales Ratio
count   1.097629e+06  1.097629e+06    1.097629e+06  1.097629e+06  1.097629e+06
mean    5.370357e+05  2.011218e+03    2.818016e+05  4.053146e+05  9.603926e+00
std     7.526074e+06  6.773485e+00    1.657890e+06  5.143492e+06  1.801664e+03
min     0.000000e+00  2.001000e+03    0.000000e+00  0.000000e+00  0.000000e+00
25%     3.071300e+04  2.005000e+03    8.909000e+04  1.450000e+05  4.778667e-01
50%     8.070600e+04  2.011000e+03    1.405800e+05  2.330000e+05  6.105663e-01
75%     1.703410e+05  2.018000e+03    2.282700e+05  3.750000e+05  7.707200e-01
max     2.000500e+09  2.022000e+03    8.815100e+08  5.000000e+09  1.226420e+06

In [6]:
df.isna().sum().sort_values(ascending=False)

OPM remarks         1084598
Assessor Remarks     926401
Location             799518
Non Use Code         784178
Residential Type     398389
Property Type        382446
Address                  51
Date Recorded             2
Serial Number             0
List Year                 0
Town                      0
Assessed Value            0
Sale Amount               0
Sales Ratio               0
dtype: int64

In [7]:
df.isnull().mean().sort_values(ascending=False)

OPM remarks         0.988128
Assessor Remarks    0.844002
Location            0.728405
Non Use Code        0.714429
Residential Type    0.362954
Property Type       0.348429
Address             0.000046
Date Recorded       0.000002
Serial Number       0.000000
List Year           0.000000
Town                0.000000
Assessed Value      0.000000
Sale Amount         0.000000
Sales Ratio         0.000000
dtype: float64

In [8]:
# Excluding OPM remarks & Assessor Remarks, because of high missing values
# Drop one or more columns by name
df = df.drop(columns=["OPM remarks", "Assessor Remarks", "Non Use Code", "Location"])

In [9]:
df

Serial Number  List Year Date Recorded           Town  \
0               220008       2022    01/30/2023        Andover   
1              2020348       2020    09/13/2021        Ansonia   
2                20002       2020    10/02/2020        Ashford   
3               210317       2021    07/05/2022           Avon   
4               200212       2020    03/09/2021           Avon   
...                ...        ...           ...            ...   
1097624         190272       2019    06/24/2020     New London   
1097625         190284       2019    11/27/2019      Waterbury   
1097626         190129       2019    04/27/2020  Windsor Locks   
1097627         190504       2019    06/03/2020     Middletown   
1097628         190344       2019    12/20/2019        Milford   

                  Address  Assessed Value  Sale Amount  Sales Ratio  \
0             618 ROUTE 6        139020.0     232000.0     0.599200   
1         230 WAKELEE AVE        150500.0     325000.0     0.463000   
2         390 TURNPIKE RD        253000.0     430000.0     0.588300   
3         53 COTSWOLD WAY        329730.0     805000.0     0.409600   
4        5 CHESTNUT DRIVE        130400.0     179900.0     0.724800   
...                   ...             ...          ...          ...   
1097624       4 BISHOP CT         60410.0      53100.0     1.137665   
1097625   126 PERKINS AVE         68280.0      76000.0     0.898400   
1097626    19 HATHAWAY ST        121450.0     210000.0     0.578300   
1097627      8 BYSTREK DR        203360.0     280000.0     0.726300   
1097628   250 RESEARCH DR       4035970.0    7450000.0     0.541700   

         Property Type Residential Type  
0          Residential    Single Family  
1           Commercial              NaN  
2          Residential    Single Family  
3          Residential    Single Family  
4          Residential            Condo  
...                ...              ...  
1097624  Single Family    Single Family  
1097625  Single Family    Single Family  
1097626  Single Family    Single Family  
1097627  Single Family    Single Family  
1097628            NaN              NaN  

[1097629 rows x 10 columns]

In [10]:
df.isnull().mean().sort_values(ascending=False)

Residential Type    0.362954
Property Type       0.348429
Address             0.000046
Date Recorded       0.000002
Serial Number       0.000000
List Year           0.000000
Town                0.000000
Assessed Value      0.000000
Sale Amount         0.000000
Sales Ratio         0.000000
dtype: float64

In [11]:
# Cleaning Dataframe
df_clean = df[(df["Sale Amount"].notna()) & (df["Town"].notna())]

# Convertir explicitement en datetime
df["Date Recorded"] = pd.to_datetime(df["Date Recorded"], errors="coerce", format="%m/%d/%Y")

#print(df["Date Recorded"].dtype)

df_clean["Date Recorded"] = pd.to_datetime(df_clean["Date Recorded"], errors='coerce')
df_clean["Year"] = df_clean["Date Recorded"].dt.year

In [12]:
# Filling Na's of cleaned dataframe with 0
df_clean["Year"] = df_clean["Year"].fillna(0)
df_clean["Property Type"] = df_clean["Property Type"].fillna('Unknown')
df_clean["Residential Type"] = df_clean["Residential Type"].fillna('Unknown')
df_clean["Address"] = df_clean["Address"].fillna('Unknown')
df_clean["Date Recorded"] = df_clean["Date Recorded"].fillna('Unknown')

In [13]:
# Example: Turn comma numbers into integers
df_clean["Year"] = df_clean["Year"].fillna(0).astype(int)

# Add Column 'State'
df_clean['State'] = 'Connecticut'

In [14]:
df_clean

Serial Number  List Year        Date Recorded           Town  \
0               220008       2022  2023-01-30 00:00:00        Andover   
1              2020348       2020  2021-09-13 00:00:00        Ansonia   
2                20002       2020  2020-10-02 00:00:00        Ashford   
3               210317       2021  2022-07-05 00:00:00           Avon   
4               200212       2020  2021-03-09 00:00:00           Avon   
...                ...        ...                  ...            ...   
1097624         190272       2019  2020-06-24 00:00:00     New London   
1097625         190284       2019  2019-11-27 00:00:00      Waterbury   
1097626         190129       2019  2020-04-27 00:00:00  Windsor Locks   
1097627         190504       2019  2020-06-03 00:00:00     Middletown   
1097628         190344       2019  2019-12-20 00:00:00        Milford   

                  Address  Assessed Value  Sale Amount  Sales Ratio  \
0             618 ROUTE 6        139020.0     232000.0     0.599200   
1         230 WAKELEE AVE        150500.0     325000.0     0.463000   
2         390 TURNPIKE RD        253000.0     430000.0     0.588300   
3         53 COTSWOLD WAY        329730.0     805000.0     0.409600   
4        5 CHESTNUT DRIVE        130400.0     179900.0     0.724800   
...                   ...             ...          ...          ...   
1097624       4 BISHOP CT         60410.0      53100.0     1.137665   
1097625   126 PERKINS AVE         68280.0      76000.0     0.898400   
1097626    19 HATHAWAY ST        121450.0     210000.0     0.578300   
1097627      8 BYSTREK DR        203360.0     280000.0     0.726300   
1097628   250 RESEARCH DR       4035970.0    7450000.0     0.541700   

         Property Type Residential Type  Year        State  
0          Residential    Single Family  2023  Connecticut  
1           Commercial          Unknown  2021  Connecticut  
2          Residential    Single Family  2020  Connecticut  
3          Residential    Single Family  2022  Connecticut  
4          Residential            Condo  2021  Connecticut  
...                ...              ...   ...          ...  
1097624  Single Family    Single Family  2020  Connecticut  
1097625  Single Family    Single Family  2019  Connecticut  
1097626  Single Family    Single Family  2020  Connecticut  
1097627  Single Family    Single Family  2020  Connecticut  
1097628        Unknown          Unknown  2019  Connecticut  

[1097629 rows x 12 columns]

In [15]:
df_clean['Serial Number'].nunique()

96220

#### Interpretation:
- we have 96220 sellers over 1097628 transactions, it means in average, each sellers is has sold with 11.41 property transactions for these 20 years.

In [16]:
df_clean['State'].unique()

array(['Connecticut'], dtype=object)

In [17]:
df_clean = df_clean.drop(columns="State")

In [18]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1097629 entries, 0 to 1097628
Data columns (total 11 columns):
 #   Column            Non-Null Count    Dtype  
---  ------            --------------    -----  
 0   Serial Number     1097629 non-null  int64  
 1   List Year         1097629 non-null  int64  
 2   Date Recorded     1097629 non-null  object 
 3   Town              1097629 non-null  object 
 4   Address           1097629 non-null  object 
 5   Assessed Value    1097629 non-null  float64
 6   Sale Amount       1097629 non-null  float64
 7   Sales Ratio       1097629 non-null  float64
 8   Property Type     1097629 non-null  object 
 9   Residential Type  1097629 non-null  object 
 10  Year              1097629 non-null  int32  
dtypes: float64(3), int32(1), int64(2), object(5)
memory usage: 87.9+ MB


In [19]:
df_clean.groupby('List Year').count()

Serial Number  Date Recorded   Town  Address  Assessed Value  \
List Year                                                                 
2001               59584          59584  59584    59584           59584   
2002               59682          59682  59682    59682           59682   
2003               64239          64239  64239    64239           64239   
2004               84056          84056  84056    84056           84056   
2005               61602          61602  61602    61602           61602   
2006               48785          48785  48785    48785           48785   
2007               35617          35617  35617    35617           35617   
2008               32735          32735  32735    32735           32735   
2009               42508          42508  42508    42508           42508   
2010               33491          33491  33491    33491           33491   
2011               31065          31065  31065    31065           31065   
2012               35973          35973  35973    35973           35973   
2013               39943          39943  39943    39943           39943   
2014               49563          49563  49563    49563           49563   
2015               46651          46651  46651    46651           46651   
2016               49773          49773  49773    49773           49773   
2017               45691          45691  45691    45691           45691   
2018               50709          50709  50709    50709           50709   
2019               58954          58954  58954    58954           58954   
2020               66592          66592  66592    66592           66592   
2021               56946          56946  56946    56946           56946   
2022               43470          43470  43470    43470           43470   

           Sale Amount  Sales Ratio  Property Type  Residential Type   Year  
List Year                                                                    
2001             59584        59584          59584             59584  59584  
2002             59682        59682          59682             59682  59682  
2003             64239        64239          64239             64239  64239  
2004             84056        84056          84056             84056  84056  
2005             61602        61602          61602             61602  61602  
2006             48785        48785          48785             48785  48785  
2007             35617        35617          35617             35617  35617  
2008             32735        32735          32735             32735  32735  
2009             42508        42508          42508             42508  42508  
2010             33491        33491          33491             33491  33491  
2011             31065        31065          31065             31065  31065  
2012             35973        35973          35973             35973  35973  
2013             39943        39943          39943             39943  39943  
2014             49563        49563          49563             49563  49563  
2015             46651        46651          46651             46651  46651  
2016             49773        49773          49773             49773  49773  
2017             45691        45691          45691             45691  45691  
2018             50709        50709          50709             50709  50709  
2019             58954        58954          58954             58954  58954  
2020             66592        66592          66592             66592  66592  
2021             56946        56946          56946             56946  56946  
2022             43470        43470          43470             43470  43470

#### Interpretation:
- The trend follows a constant number of transaction and even raises since 2001
- We can see at the year 2007 and 2008, the subprime crisis drop the transactions we drop from close to 50k to 35k and 32k ~ 15k of differences
- In 2009, it increases again and decreases again and in 2012, it started to stabilize before reaching the same amount of transactions to 2006.
- In 2020 to 2022, it decreases again. It may due to the covid-19, because the interest rate is high, so we have fewer people who would make loan because it is not worth it.

In [20]:
df_clean.groupby('List Year').count().sort_values("Sale Amount", ascending=False)

Serial Number  Date Recorded   Town  Address  Assessed Value  \
List Year                                                                 
2004               84056          84056  84056    84056           84056   
2020               66592          66592  66592    66592           66592   
2003               64239          64239  64239    64239           64239   
2005               61602          61602  61602    61602           61602   
2002               59682          59682  59682    59682           59682   
2001               59584          59584  59584    59584           59584   
2019               58954          58954  58954    58954           58954   
2021               56946          56946  56946    56946           56946   
2018               50709          50709  50709    50709           50709   
2016               49773          49773  49773    49773           49773   
2014               49563          49563  49563    49563           49563   
2006               48785          48785  48785    48785           48785   
2015               46651          46651  46651    46651           46651   
2017               45691          45691  45691    45691           45691   
2022               43470          43470  43470    43470           43470   
2009               42508          42508  42508    42508           42508   
2013               39943          39943  39943    39943           39943   
2012               35973          35973  35973    35973           35973   
2007               35617          35617  35617    35617           35617   
2010               33491          33491  33491    33491           33491   
2008               32735          32735  32735    32735           32735   
2011               31065          31065  31065    31065           31065   

           Sale Amount  Sales Ratio  Property Type  Residential Type   Year  
List Year                                                                    
2004             84056        84056          84056             84056  84056  
2020             66592        66592          66592             66592  66592  
2003             64239        64239          64239             64239  64239  
2005             61602        61602          61602             61602  61602  
2002             59682        59682          59682             59682  59682  
2001             59584        59584          59584             59584  59584  
2019             58954        58954          58954             58954  58954  
2021             56946        56946          56946             56946  56946  
2018             50709        50709          50709             50709  50709  
2016             49773        49773          49773             49773  49773  
2014             49563        49563          49563             49563  49563  
2006             48785        48785          48785             48785  48785  
2015             46651        46651          46651             46651  46651  
2017             45691        45691          45691             45691  45691  
2022             43470        43470          43470             43470  43470  
2009             42508        42508          42508             42508  42508  
2013             39943        39943          39943             39943  39943  
2012             35973        35973          35973             35973  35973  
2007             35617        35617          35617             35617  35617  
2010             33491        33491          33491             33491  33491  
2008             32735        32735          32735             32735  32735  
2011             31065        31065          31065             31065  31065

#### Interpretation:
- Top 3 by year that we had the most transactions: 2004, 2020 and 2003. Than the rest is pretty constant after 2005
  - 2003: real estate boom -> between 2001 and 2006
  - 2004: real estate boom (Easy credit, low interest rates, issuing risky mortgages to unqualified buyers (called subprime loans))
  - 2020: interest rate low, remote work (people move to countryside, so the transactions of houses increase), recovery plan, etc.
- Top 3 by year that we had the less transactions: 2011, 2008 and 2010. All the worst is pretty constant
  - 2007-2008: we had the subprime crisis, so it makes sens.
  - 2009-2010: we had the re-activity
  - 2010-2011: we had the waiting period which people wasn't confident to sell or buy and a lot of transactions was postponed
  - 2011: It fails so in 2011, it decreases.
- So, **economics activity** (**credit**, **low interest rates**, **mortgages**, **crisis**) are the indicator we should look closely to have more transactions

In [21]:
df_clean.isna().sum()

Serial Number       0
List Year           0
Date Recorded       0
Town                0
Address             0
Assessed Value      0
Sale Amount         0
Sales Ratio         0
Property Type       0
Residential Type    0
Year                0
dtype: int64

In [22]:
df_clean["Good Deal"] = ( df_clean['Sale Amount'] > df_clean['Assessed Value']).astype(int)

In [23]:
df_clean.head()

Serial Number  List Year        Date Recorded     Town           Address  \
0         220008       2022  2023-01-30 00:00:00  Andover       618 ROUTE 6   
1        2020348       2020  2021-09-13 00:00:00  Ansonia   230 WAKELEE AVE   
2          20002       2020  2020-10-02 00:00:00  Ashford   390 TURNPIKE RD   
3         210317       2021  2022-07-05 00:00:00     Avon   53 COTSWOLD WAY   
4         200212       2020  2021-03-09 00:00:00     Avon  5 CHESTNUT DRIVE   

   Assessed Value  Sale Amount  Sales Ratio Property Type Residential Type  \
0        139020.0     232000.0       0.5992   Residential    Single Family   
1        150500.0     325000.0       0.4630    Commercial          Unknown   
2        253000.0     430000.0       0.5883   Residential    Single Family   
3        329730.0     805000.0       0.4096   Residential    Single Family   
4        130400.0     179900.0       0.7248   Residential            Condo   

   Year  Good Deal  
0  2023          1  
1  2021          1  
2  2020          1  
3  2022          1  
4  2021          1

In [24]:
df_clean.groupby('Good Deal').count()

Serial Number  List Year  Date Recorded    Town  Address  \
Good Deal                                                             
0                 138557     138557         138557  138557   138557   
1                 959072     959072         959072  959072   959072   

           Assessed Value  Sale Amount  Sales Ratio  Property Type  \
Good Deal                                                            
0                  138557       138557       138557         138557   
1                  959072       959072       959072         959072   

           Residential Type    Year  
Good Deal                            
0                    138557  138557  
1                    959072  959072

In [25]:
df_clean = df_clean[df_clean["Town"] != "***Unknown***"]

#### Interpretation
- We have more good deal than bad deal

In [26]:
df_clean.groupby(['Good Deal', 'Town']).count().sort_values("Sale Amount", ascending=False)

Serial Number  List Year  Date Recorded  Address  \
Good Deal Town                                                              
1         Stamford               33181      33181          33181    33181   
          Bridgeport             27830      27830          27830    27830   
          Waterbury              25043      25043          25043    25043   
          Norwalk                24525      24525          24525    24525   
          West Hartford          21100      21100          21100    21100   
...                                ...        ...            ...      ...   
0         Warren                    84         84             84       84   
          Bozrah                    80         80             80       80   
          Scotland                  80         80             80       80   
          Eastford                  78         78             78       78   
          Union                     57         57             57       57   

                         Assessed Value  Sale Amount  Sales Ratio  \
Good Deal Town                                                      
1         Stamford                33181        33181        33181   
          Bridgeport              27830        27830        27830   
          Waterbury               25043        25043        25043   
          Norwalk                 24525        24525        24525   
          West Hartford           21100        21100        21100   
...                                 ...          ...          ...   
0         Warren                     84           84           84   
          Bozrah                     80           80           80   
          Scotland                   80           80           80   
          Eastford                   78           78           78   
          Union                      57           57           57   

                         Property Type  Residential Type   Year  
Good Deal Town                                                   
1         Stamford               33181             33181  33181  
          Bridgeport             27830             27830  27830  
          Waterbury              25043             25043  25043  
          Norwalk                24525             24525  24525  
          West Hartford          21100             21100  21100  
...                                ...               ...    ...  
0         Warren                    84                84     84  
          Bozrah                    80                80     80  
          Scotland                  80                80     80  
          Eastford                  78                78     78  
          Union                     57                57     57  

[338 rows x 10 columns]

#### Interpretation
- Top 5 town where they have the best deal:
  - Stamford
  - Bridgeport
  - Waterbury
  - Norwalk
  - West Hartford
- Top 5 town where they have the worst deal:
  - Warren
  - Bozrah
  - Scotland
  - Eastford
  - Union

In [27]:
df_clean_gd = df_clean[df_clean["Good Deal"] == 1]
df_clean_bd = df_clean[df_clean["Good Deal"] == 0]

In [28]:
df_clean_gd.groupby(['List Year', 'Town']).count().sort_values(by=["List Year", "Sale Amount"], ascending=[False, False])

Serial Number  Date Recorded  Address  Assessed Value  \
List Year Town                                                                
2022      Waterbury            1674           1674     1674            1674   
          Stamford             1539           1539     1539            1539   
          Bridgeport           1536           1536     1536            1536   
          Norwalk              1214           1214     1214            1214   
          Hamden                956            956      956             956   
...                             ...            ...      ...             ...   
2001      Hartland               33             33       33              33   
          Bozrah                 31             31       31              31   
          Canaan                 30             30       30              30   
          Union                  25             25       25              25   
          New Haven               2              2        2               2   

                      Sale Amount  Sales Ratio  Property Type  \
List Year Town                                                  
2022      Waterbury          1674         1674           1674   
          Stamford           1539         1539           1539   
          Bridgeport         1536         1536           1536   
          Norwalk            1214         1214           1214   
          Hamden              956          956            956   
...                           ...          ...            ...   
2001      Hartland             33           33             33   
          Bozrah               31           31             31   
          Canaan               30           30             30   
          Union                25           25             25   
          New Haven             2            2              2   

                      Residential Type  Year  Good Deal  
List Year Town                                           
2022      Waterbury               1674  1674       1674  
          Stamford                1539  1539       1539  
          Bridgeport              1536  1536       1536  
          Norwalk                 1214  1214       1214  
          Hamden                   956   956        956  
...                                ...   ...        ...  
2001      Hartland                  33    33         33  
          Bozrah                    31    31         31  
          Canaan                    30    30         30  
          Union                     25    25         25  
          New Haven                  2     2          2  

[3061 rows x 10 columns]

In [29]:
df_clean_bd.groupby(['List Year', 'Town']).count().sort_values(by=["List Year", "Sale Amount"], ascending=[False, False])

Serial Number  Date Recorded  Address  Assessed Value  \
List Year Town                                                                 
2022      Bridgeport             136            136      136             136   
          Waterbury               97             97       97              97   
          Stamford                96             96       96              96   
          Norwich                 83             83       83              83   
          Greenwich               81             81       81              81   
...                              ...            ...      ...             ...   
2001      Canaan                   2              2        2               2   
          Chaplin                  2              2        2               2   
          Colebrook                2              2        2               2   
          Roxbury                  2              2        2               2   
          East Granby              1              1        1               1   

                       Sale Amount  Sales Ratio  Property Type  \
List Year Town                                                   
2022      Bridgeport           136          136            136   
          Waterbury             97           97             97   
          Stamford              96           96             96   
          Norwich               83           83             83   
          Greenwich             81           81             81   
...                            ...          ...            ...   
2001      Canaan                 2            2              2   
          Chaplin                2            2              2   
          Colebrook              2            2              2   
          Roxbury                2            2              2   
          East Granby            1            1              1   

                       Residential Type  Year  Good Deal  
List Year Town                                            
2022      Bridgeport                136   136        136  
          Waterbury                  97    97         97  
          Stamford                   96    96         96  
          Norwich                    83    83         83  
          Greenwich                  81    81         81  
...                                 ...   ...        ...  
2001      Canaan                      2     2          2  
          Chaplin                     2     2          2  
          Colebrook                   2     2          2  
          Roxbury                     2     2          2  
          East Granby                 1     1          1  

[3021 rows x 10 columns]

In [31]:
df_clean_bd.groupby('List Year').count().sort_values(by="Serial Number", ascending=False)

Serial Number  Date Recorded   Town  Address  Assessed Value  \
List Year                                                                 
2014               10245          10245  10245    10245           10245   
2009               10235          10235  10235    10235           10235   
2011               10143          10143  10143    10143           10143   
2010                9935           9935   9935     9935            9935   
2015                8941           8941   8941     8941            8941   
2013                8661           8661   8661     8661            8661   
2012                8318           8318   8318     8318            8318   
2016                8197           8197   8197     8197            8197   
2008                6595           6595   6595     6595            6595   
2017                6405           6405   6405     6405            6405   
2018                6307           6307   6307     6307            6307   
2019                5762           5762   5762     5762            5762   
2004                5579           5579   5579     5579            5579   
2001                5162           5162   5162     5162            5162   
2002                4538           4538   4538     4538            4538   
2003                4199           4199   4199     4199            4199   
2020                3920           3920   3920     3920            3920   
2007                3657           3657   3657     3657            3657   
2006                3262           3262   3262     3262            3262   
2005                3211           3211   3211     3211            3211   
2021                2836           2836   2836     2836            2836   
2022                2449           2449   2449     2449            2449   

           Sale Amount  Sales Ratio  Property Type  Residential Type   Year  \
List Year                                                                     
2014             10245        10245          10245             10245  10245   
2009             10235        10235          10235             10235  10235   
2011             10143        10143          10143             10143  10143   
2010              9935         9935           9935              9935   9935   
2015              8941         8941           8941              8941   8941   
2013              8661         8661           8661              8661   8661   
2012              8318         8318           8318              8318   8318   
2016              8197         8197           8197              8197   8197   
2008              6595         6595           6595              6595   6595   
2017              6405         6405           6405              6405   6405   
2018              6307         6307           6307              6307   6307   
2019              5762         5762           5762              5762   5762   
2004              5579         5579           5579              5579   5579   
2001              5162         5162           5162              5162   5162   
2002              4538         4538           4538              4538   4538   
2003              4199         4199           4199              4199   4199   
2020              3920         3920           3920              3920   3920   
2007              3657         3657           3657              3657   3657   
2006              3262         3262           3262              3262   3262   
2005              3211         3211           3211              3211   3211   
2021              2836         2836           2836              2836   2836   
2022              2449         2449           2449              2449   2449   

           Good Deal  
List Year             
2014           10245  
2009           10235  
2011           10143  
2010            9935  
2015            8941  
2013            8661  
2012            8318  
2016            8197  
2008            6595  
2017            6405  
2018            6307  
2019          